base/iobuffer.jl

```diff
@@ -131,17 +210,26 @@ function IOBuffer(;
         append::Union{Bool,Nothing}=nothing,
         truncate::Union{Bool,Nothing}=true,
         maxsize::Integer=typemax(Int),
-        sizehint::Union{Integer,Nothing}=nothing)
-    size = sizehint !== nothing ? Int(sizehint) : maxsize != typemax(Int) ? Int(maxsize) : 32
+        sizehint::Union{Integer,Nothing}=nothing,
+    )
+    mz = Int(maxsize)::Int
+    if mz < 0
+        throw(ArgumentError("negative maxsize"))
+    end
+    size = if sizehint !== nothing
+        # Allow negative sizehint, just like `sizehint!` does
+        min(mz, max(0, Int(sizehint)::Int))
+    else
+        min(mz, 32)
+    end
     flags = open_flags(read=read, write=write, append=append, truncate=truncate)
-    buf = IOBuffer(
-        StringMemory(size),
-        read=flags.read,
-        write=flags.write,
-        append=flags.append,
-        truncate=flags.truncate,
-        maxsize=maxsize)
-    fill!(buf.data, 0)
+    # A common usecase of IOBuffer is to incrementally construct strings. By using StringMemory
+    # as the default storage, we can turn the result into a string without copying.
+    data = fill!(StringMemory(size), 0)
+    buf = GenericIOBuffer{Memory{UInt8}}(unsafe_method, data, flags.read, flags.write, true, flags.append, mz)
+    if flags.truncate
+        buf.size = 0
+    end
```

Comment:
Since there is no initial data, I think it makes sense to always set `size` to zero.

If `size` is zero, you can probably remove the `data = fill!(StringMemory(size), 0)`.

Otherwise, passing `truncate=false` exposes the uninitialized memory for reading.

```julia-repl
julia> read(IOBuffer(;maxsize=34, truncate=false))
34-element Vector{UInt8}:
 0x00
 0x00
 0x00
 0x00

```